File: bff/src/controllers/productController.js
// const productService = require("../services/productService");

// const getProducts = async (req, res) => {
//   try {
//     const products = await productService.getProducts();
//     res.json(products);
//   } catch (error) {
//     res.status(500).json({ message: "Error fetching products", error });
//   }
// };

// const getProductById = async (req, res) => {
//   try {
//     const { productId } = req.params;
//     const product = await productService.getProductById(productId);
//     res.json(product);
//   } catch (error) {
//     res.status(500).json({ message: "Error fetching product", error });
//   }
// };

// const createProduct = async (req, res) => {
//   try {
//     const { file, product } = req.body;
//     const newProduct = await productService.createProduct(file, product);
//     res.json(newProduct);
//   } catch (error) {
//     res.status(500).json({ message: "Error creating product", error });
//   }
// };

// const getProductsByCategoryId = async (req, res) => {
//   try {
//     const { categoryId } = req.params;
//     const products = await productService.getProductsByCategoryId(categoryId);
//     res.json(products);
//   } catch (error) {
//     console.error("Error fetching products by category:", error);
//     res.status(500).json({
//       message: "Error fetching products by category",
//       error: error.message,
//     });
//   }
// };

// module.exports = {
//   getProducts,
//   getProductById,
//   createProduct,
//   getProductsByCategoryId,
// };

const productService = require("../services/productService");

const getProducts = async (req, res) => {
  try {
    const {
      categoryId,
      sortBy,
      minPrice,
      maxPrice,
      search,
      page = 0,
      size = 4,
    } = req.query;
    const products = await productService.getProducts(
      categoryId,
      sortBy,
      minPrice,
      maxPrice,
      search,
      page,
      size
    );
    // console.log(products);
    res.json(products);
  } catch (error) {
    res.status(500).json({ message: "Error fetching products", error });
  }
};

const getProductById = async (req, res) => {
  try {
    const { productId } = req.query;
    const product = await productService.getProductById(productId);
    if (product) {
      res.json(product);
    } else {
      res.status(404).json({ message: "Product not found" });
    }
  } catch (error) {
    res.status(500).json({ message: "Error fetching product", error });
  }
};

const createProduct = async (req, res) => {
  try {
    const { file } = req.files;
    const product = req.body.product;
    const newProduct = await productService.createProduct(file, product);
    res.json(newProduct);
  } catch (error) {
    res.status(500).json({ message: "Error creating product", error });
  }
};

module.exports = {
  getProducts,
  getProductById,
  createProduct,
};

File: backend/service-product/src/main/java/com/onlinestore/serviceproduct/controller/ProductController.java
package com.onlinestore.serviceproduct.controller;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.onlinestore.serviceproduct.dto.ProductDTO;
import com.onlinestore.serviceproduct.service.ProductService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;

@RestController
@RequestMapping("api/v1/products")
@CrossOrigin("http://localhost:4000")
public class ProductController {

    @Autowired
    private ProductService productService;

    @PostMapping("/addProduct")
    public ResponseEntity<ProductDTO> uploadImage(
            @RequestParam("file") MultipartFile file,
            @RequestParam("product") String productDTOString) {
        try {
            ObjectMapper objectMapper = new ObjectMapper();
            ProductDTO productDTO = objectMapper.readValue(productDTOString, ProductDTO.class);
            ProductDTO savedProduct = productService.saveProductWithImage(file, productDTO);
            return ResponseEntity.ok(savedProduct);
        } catch (IOException e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

    @GetMapping
    public ResponseEntity<Page<ProductDTO>> getProducts(
            @RequestParam(required = false) String categoryId,
            @RequestParam(required = false) String sortBy,
            @RequestParam(required = false) Double minPrice,
            @RequestParam(required = false) Double maxPrice,
            @RequestParam(required = false) String search,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "4") int size) {
        try {
            Pageable pageable = PageRequest.of(page, size, createSort(sortBy));
            Page<ProductDTO> products = productService.getFilteredProducts(categoryId, minPrice, maxPrice, search, pageable);
            return ResponseEntity.ok(products);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

    @GetMapping("/getProductById")
    public ResponseEntity<ProductDTO> getProductById(@RequestParam String productId) {
        return productService.getProductById(productId)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    private Sort createSort(String sortBy) {
    if (sortBy == null || sortBy.isEmpty()) {
        return Sort.unsorted();
    }
    switch (sortBy) {
        case "priceAsc":
            return Sort.by(Sort.Direction.ASC, "price");
        case "priceDesc":
            return Sort.by(Sort.Direction.DESC, "price");
        default:
            return Sort.unsorted();
    }
}
}